include/seastar/parquet/column_chunk_reader.hh

```diff
@@ -0,0 +1,396 @@
+    void reset(std::basic_string_view<uint8_t> buf, format::Encoding::type encoding);
+    // Read a batch of n values (the last batch may be smaller than n).
+    size_t read_batch(size_t n, output_type out[]);
+};
+
+// The core low-level interface. Takes the relevant metadata and an input_stream set to the beginning of a column chunk
+// and extracts batches of (repetition level, definition level, value (optional)) from it.
+template<format::Type::type T>
+class column_chunk_reader {
+public:
+    using output_type = typename value_decoder_traits<T>::output_type;
+    const schema::raw_node& _schema_node;
+private:
+    page_reader _source;
+    decompressor _decompressor;
+    level_decoder _rep_decoder;
+    level_decoder _def_decoder;
+    value_decoder<T> _val_decoder;
+    std::optional<std::vector<output_type>> _dict;
+    bool _initialized = false;
+    bool _eof = false;
+    int64_t _page_ordinal = -1; // Only used for error reporting.
+private:
+    seastar::future<> load_next_page();
+    void load_dictionary_page(page p);
+    void load_data_page(page p);
+    void load_data_page_v2(page p);
+
+    template<typename LevelT>
+    seastar::future<size_t> read_batch_internal(size_t n, LevelT def[], LevelT rep[], output_type val[]);
+public:
+    explicit column_chunk_reader(
+            const schema::raw_node& schema_node,
+            page_reader&& source,
+            format::CompressionCodec::type codec)
+        : _schema_node{schema_node}
+        , _source{std::move(source)}
+        , _decompressor{codec}
+        , _rep_decoder{schema_node.rep_level}
+        , _def_decoder{schema_node.def_level}
+        , _val_decoder{
+            _schema_node.info.__isset.type_length
+            ? std::optional<uint32_t>{static_cast<uint32_t>(_schema_node.info.type_length)}
+            : std::optional<uint32_t>{}}
+        {};
+    // Read a batch of n (rep, def, value) triplets. The last batch may be smaller than n.
+    // Return the number of triplets read. Note that null values are not read into the output array.
+    // Example output: def == [1, 1, 0, 1, 0], rep = [0, 0, 0, 0, 0], val = ["a", "b", "d"].
+    template<typename LevelT>
+    seastar::future<size_t> read_batch(size_t n, LevelT def[], LevelT rep[], output_type val[]);
+};
+
+template<format::Type::type T>
+template<typename LevelT>
+seastar::future<size_t>
+column_chunk_reader<T>::read_batch_internal(size_t n, LevelT def[], LevelT rep[], output_type val[]) {
+    if (_eof) {
+        return seastar::make_ready_future<size_t>(0);
+    }
+    if (!_initialized) {
+        return load_next_page().then([this, n, def, rep, val] {
+            return read_batch_internal(n, def, rep, val);
+        });
+    }
+    size_t def_levels_read = _def_decoder.read_batch(n, def);
+    size_t rep_levels_read = _rep_decoder.read_batch(n, rep);
+    if (def_levels_read != rep_levels_read) {
+        throw parquet_exception::corrupted_file(seastar::format(
```

Comment:
It's more efficient to directly create an exceptional future instead of throwing:
```cpp
    return make_exception_future<size_t>(parquet_exception(...etc));
```
That's because throwing exceptions is a bottleneck, especially in a shared-nothing application - global locks may be involved, depending on the C++ library version.

This comment applies to other throws as well. It's generally fine to throw, but if we're in top-level function that returns a future, it's better to avoid the throw-catch ritual and simply create the future directly.